### PR TITLE
list of contracts, related to #286

### PR DESCRIPTION
**Pull request by bergauz**
Screenshot 2019-09-24 at 04 14 29

### CODE REVIEW

**Comment by jmank88** [comment]
Awesome!

What does it look like on mobile?

I wonder how this column ordering would look:
` # Symbol Name Type Txs Internal-txs Holders Token-txs Address`

That row 3 is already pretty cramped with 5 interfaces - we should plan for handling even more smoothly. Since the `Type` column isn't a single sortable value like the others, what if it wasn't a column, and the types were listed underneath the other info from left to right? Looking at the code, I see this is a table, so maybe not a feasible tweak right now.

File: server/backend/db_backend.go [hunk 1]
```diff
@@ -603,6 +606,47 @@ func (self *MongoBackend) updateContract(contract *models.Contract) bool {
 	return true
 }
 
+func (self *MongoBackend) getContracts(filter *models.ContractsFilter) []*models.Address {
+	var addresses []*models.Address
+	var sortQuery string
+	findQuery := bson.M{"contract": true}
+	if filter.TokenName != "" {
+		findQuery["token_name"] = bson.RegEx{filter.TokenName, "i"}
```
Comment (jmank88) [inline server/backend/db_backend.go]: Are we allowing users to submit arbitrary regex to the database? That seems exploitable.

**Comment by bergauz** [comment]
Screenshot 2019-09-25 at 00 14 07
Screenshot 2019-09-25 at 00 14 22

**Comment by bergauz** [comment]
about table if columns gonna increase we can replace with icons like in gochain stats

**Comment by jmank88** [comment]
That looks good. I still think the ordering can be improved, one of these would make sense in terms of importance to the user, and alignment in the table columns:
` # Symbol Name Type Txs Internal-txs Holders Token-txs Address`

**Comment by bergauz** [comment]
Screenshot 2019-09-26 at 14 47 58

**Comment by jmank88** [comment]
Looks good! Needs some rebase/merge cleanup though it looks like.

File: server/backend/db_backend.go [hunk 2]
```diff
@@ -605,6 +609,47 @@ func (self *MongoBackend) updateContract(contract *models.Contract) bool {
...
+	}
+	err := self.mongo.
+		C("Addresses").
+		Find(findQuery).
+		Sort(sortQuery).
+		Skip(filter.Skip).
+		Limit(filter.Limit).
+		All(&addresses)
+	if err != nil {
+		log.Debug().Err(err).Msg("GetContracts")
```
Comment (jmank88) [inline server/backend/db_backend.go]: This is the trouble. Need to pass in the new logger type, and this is probably ERROR level, isn't it?
```suggestion
		lgr.Error("Failed to query contracts", zap.Error(err))
```

Comment (bergauz) [reply server/backend/db_backend.go]: @jmank88  yes, fixed

File: server/models/address.go [hunk 3]
```diff
@@ -27,5 +27,5 @@ type Address struct {
 }
 
 type AddressesList struct {
-	Adresses []*Address `json:"adresses"`
+	Addresses []*Address `json:"addresses"`
```
Comment (jmank88) [inline server/models/address.go]: I checked to see if this was in the public API, but there aren't actually any usages anywhere. Can this struct be removed?

File: server/backend/db_backend.go [hunk 4]
```diff
@@ -605,6 +609,47 @@ func (self *MongoBackend) updateContract(contract *models.Contract) bool {
...
+	}
+	err := self.mongo.
+		C("Addresses").
+		Find(findQuery).
+		Sort(sortQuery).
+		Skip(filter.Skip).
+		Limit(filter.Limit).
+		All(&addresses)
+	if err != nil {
+		self.Lgr.Debug("GetContracts")
```
Comment (jmank88) [inline server/backend/db_backend.go]: Isn't this error important? Also, this logging library includes `caller: <package>/<file>:<linenumber>` automatically, so descriptive messages are more helpful:
```suggestion
		self.Lgr.Error("Failed to query contracts", zap.Error(err))
```

**Comment by r-gochain** [comment]
@bergauz could you resolve the conflicts?

File: server/backend/db_backend.go [hunk 5]
```diff
@@ -622,6 +626,47 @@ func (self *MongoBackend) updateContract(contract *models.Contract) error {
 	return nil
 }
 
+func (self *MongoBackend) getContracts(filter *models.ContractsFilter) []*models.Address {
+	var addresses []*models.Address
+	var sortQuery string
+	findQuery := bson.M{"contract": true}
```
Comment (r-gochain) [inline server/backend/db_backend.go]: add index {contract:1}

**Pull request merged**